src/core/service.c

```diff
@@ -3520,21 +3525,24 @@ static int service_dispatch_exec_io(sd_event_source *source, int fd, uint32_t ev
          * the pipe to be closed (for example, a simple exit()). To deal with that we'll ignore EOFs on the pipe unless
          * the child signalled us first that it is about to call the execve(). It does so by sending us a simple
          * non-zero byte via the pipe. We also provide the child with a way to inform us in case execve() failed: if it
-         * sends a zero byte we'll ignore POLLHUP on the fd again. */
+         * sends a zero byte we'll ignore POLLHUP on the fd again.
+         * For exec, dbus, notify and notify-reload types we also get a timestamp from sd-executor, taken immediately
+         * before the target executable is execve'd, so that we can accurately track when control is handed over to the
+         * payload. */
 
         for (;;) {
-                uint8_t x;
+                uint64_t x = 0;
                 ssize_t n;
 
-                n = read(fd, &x, sizeof(x));
+                n = loop_read(fd, &x, sizeof(x), /* do_poll= */ false);
```

Comment:
I was asked to check the value instead of the size here: https://github.com/systemd/systemd/pull/32380#discussion_r1574407735

Reply:
Well, yeah, I saw that comment too. But I think the current implementation is busted.

Reply:
I have updated it so that we still check the value of the message, but also double-check that we got a full usec_t before using it. This should hopefully satisfy you both.

Reply:
Also note that this is mostly extra cautiousness, as the executor is pinned by fd exactly to ensure the protocols are always in sync, and we can't get a half read from a pipe. So in practice before the update we always read 0 or 1, and after the update we always read 0 or timestamp.